lib/pbench/server/filetree.py

```diff
@@ -298,17 +312,34 @@ def unpack(self, incoming: Path, results: Path):
         ensured by calling this indirectly through the Controller class unpack
         method.
 
-        TODO: This is a prototype for testing, and doesn't actually unpack
-        the tarball as we're going to be relying on the current unpack pipeline
-        command for some time.
-
         Args:
             incoming: Controller's directory in the INCOMING tree
             results: Controller's directory in the RESULTS tree
+
+        Raises:
+            TarballUnpackError if an exception/failure occurs while unpacking
+            the tarball.
         """
-        self.logger.warning("Tarball unpack is not yet implemented")
         unpacked = incoming / self.name
-        unpacked.mkdir(parents=True)  # Just create an empty directory for now
+        unpacked.mkdir(parents=True)
+
+        try:
+            process = subprocess.run(
+                ["tar", "-xf", self.tarball_path],
+                cwd=unpacked,
+                stdin=subprocess.DEVNULL,
+                capture_output=True,
+                text=True,
+            )
```

Comment:
Sadly, we want this to be a bit more complex, after the discussion at scrum.
1. Unpack the tarball into a "scratch" directory, which won't be found by the normal tools. The bash script used what here might amount to `incoming / f"{self.name}.unpack"`
2. After a successful unpack:
  A. Add `rx` mode for user/group/other for all directories in the unpacked subtree
  B. Add `r` mode for all non-directory files in the unpacked subtree.
  C. Rename the `{self.name}.unpack` directory to `{self.name}` to "publish" it.

Look back at the original bash script to see how this is done.

You could do it using a Python directory tree traversal. You could use `subprocess.run` for the two commands used in the bash script (a `find` and a `chmod -R`), or we can work out a single `find` command that will do both.

E.g., while I haven't debugged this, something like

```bash
find ${incoming}.unpack/ \( -type d -exec chmod ugo+rx "{}" \; \) -o \( -type f -exec chmod ugo+r "{}" \; \)
```

It might be nice to avoid the fork/exec for each file in the tree, however I'm not quite sure how to make that work with the alternation of the search pattern. We can't use the `-print0` mechanism in the bash script version because that loses the distinction between the two branches, and I'm not sure how `\+` interacts with quoting the `"{}"` to protect against files with embedded spaces. Maybe someone else has an idea.

Reply:
The `find(1)` man page seems to recommend placing the `{}` in single quotes.  However, I think that most of the concern is centered around piping the output of `find` to other tools, like `xargs`.  If we instead use `-exec`, we seem to be shielded from the issues:
```
$ find . -maxdepth 1 -type f -exec wc -l '{}' \;
0 ./one 1
0 ./two 2
0 ./??? [12]
```
In this example, `wc` has no problem recognizing the weird filenames, and it works the same with or without the quotes.  Presumably, this is because there is no shell involved, and `find` decides what the arguments to `exec*(2)` are.

And, if you change the `\;` to `\+`, it works fine, as well:
```
$ find . -maxdepth 1 -type f -exec wc -l {} \+
0 ./one 1
0 ./two 2
0 ./??? [12]
0 total
```
(Note the additional line for the total.)

Reply:
> And, if you change the \; to \+, it works fine, as well:

Yes, but what I was concerned about (and didn't go to great lengths to try) was the interaction of the `\+` across the two separate terms in my expression. Does `find` resolve the entire tree into two lists and execute just two `chmod` commands, or does it collect a list until it finds the alternate subexpression, execute it, and begin collecting a new list? Actually, your `wc -l` is a better test case than the `ls` I used, which doesn't give a way to identify a group in the output.

So I tried this:

```bash
find . \( -type d -exec du -hc '{}' \+ \) -o \( -type f -exec wc -l '{}' \+ \)
```

My test output shows that it does indeed traverse the entire tree and generate exactly two commands, a `du -hc` with all the directories and a `wc -l` with all the regular files. That's pretty cool, and exactly what we'd want. (Except, of course, for the `du` and `wc`, which wouldn't actually help us. 😆 )

Reply:
> generate exactly two commands

Actually, what I believe `find` does is accumulate arguments until the command line is "full" and then issue the command.  So, potentially, we would see more than two command invocations if the file tree were large enough.  But, yeah.

Reply:
Well, yeah, if the tree is big enough; according to the documentation, find `-exec` "works like xargs", which is what we'd hope and expect. That wasn't what I meant. I was concerned that, for example in a directory tree with a few files in each of many directories we might end up with a few files, a directory or two, another few files, and so forth. (Note, according to the documentation, there's `-execdir` which *does* behave that way, operating only across files within each directory.)